config/persistence.go

```diff
@@ -72,70 +72,124 @@ func JSONToMap(jsonData []byte) (map[string]interface{}, error) {
 		return nil, err
 	}
 
-	flatten(loaded, loaded, "")
-	return loaded, nil
+	return Flatten(loaded), nil
 }
 
-func flatten(rootMap, subMap map[string]interface{}, subKey string) {
+// Flatten returns a flattened copy of the given hierarchical config.
+func Flatten(config map[string]interface{}) (flattenedConfig map[string]interface{}) {
+	flattenedConfig = make(map[string]interface{})
+	flattenMap(flattenedConfig, config, "")
+	return flattenedConfig
+}
+
+func flattenMap(rootMap, subMap map[string]interface{}, subKey string) {
 	for key, entry := range subMap {
 
 		// get next level key
-		subbedKey := key
-		if subKey != "" {
-			subbedKey = fmt.Sprintf("%s/%s", subKey, key)
-		}
+		subbedKey := path.Join(subKey, key)
 
 		// check for next subMap
 		nextSub, ok := entry.(map[string]interface{})
 		if ok {
-			flatten(rootMap, nextSub, subbedKey)
-			delete(rootMap, key)
-		} else if subKey != "" {
+			flattenMap(rootMap, nextSub, subbedKey)
+		} else {
 			// only set if not on root level
 			rootMap[subbedKey] = entry
 		}
 	}
 }
 
-// MapToJSON expands a flattened map and returns it as json. The map is altered in the process.
-func MapToJSON(values map[string]interface{}) ([]byte, error) {
-	expand(values)
-	return json.MarshalIndent(values, "", "  ")
+// MapToJSON expands a flattened map and returns it as json.
+func MapToJSON(config map[string]interface{}) ([]byte, error) {
+	return json.MarshalIndent(Expand(config), "", "  ")
+}
+
+// Expand returns a hierarchical copy of the given flattened config.
+func Expand(flattenedConfig map[string]interface{}) (config map[string]interface{}) {
+	config = make(map[string]interface{})
+	for key, entry := range flattenedConfig {
+		PutValueIntoHierarchicalConfig(config, key, entry)
+	}
+	return config
+}
+
+// PutValueIntoHierarchicalConfig injects a configuration entry into an hierarchical config map. Conflicting entries will be replaced.
+func PutValueIntoHierarchicalConfig(config map[string]interface{}, key string, value interface{}) {
+	parts := strings.Split(key, "/")
+
+	// create/check maps for all parts except the last one
+	subMap := config
+	for i := 0; i < len(parts)-1; i++ {
```

Comment:
Is there a reason to use `for i := 0; i< ...` instead of `for _, part := range parts` ?